### ISSUE TITLE
void EntityManager.delete(Object primaryKey)

### ISSUE THREAD

**Issue by lukasj** [eclipse-ee4j/jpa-api#147]
According to the current JPA specification, deleting object is possible by the following ways:

*   em.remove(em.find(pk)) – Bad: First loads an instance from the DB into the cache, just for the sake of deleting it.
*   em.createNativeQuery(sql) – Bad: SQL string has to be parsed at runtime, and can contain typos not detectable at compile time.
*   em.createQuery(jpql) – Bad: JPDQ string has to be parsed at runtime, and can contain typos not detectable at compile time.
*   em.createQuery(CriteriaDelete) – Bad: Rather complex java code to write even for simple cases.

Hence I would like to propose the addition of one method:

```
void EntityManager.delete(Object pk) throws PersistenceException
```

This method directly issues SQL DELETE WHERE id = ? command to the database. If the JDBC driver answers with "zero rows affected statement", then EntityNotFoundException is thrown. Otherwise (i. e. "one row affected" is reported by the JDBC driver) the object is evicted from the cache.

Benefits:
+ Very simple to write and super-easy to understand even for beginners
+ Very fast to execute
+ Safe against typos
+ Typesafe even for compound keys
+ No useless SQL SELECT roundtrip
+ Correctly reports whether there was a problem, whether there was success, or whether there was nothing to delete.

From the aspect of API completeness, this method is really missing since the beginning of JPA, so IMHO it fits perfectly into a maintenance release.
#### Affected Versions
[2.2]

**Comment by mkarg** [comment]
@mkarg I was the original reporter.

**Comment by maffe** [comment]
> * `em.remove(em.find(pk))` – Bad: First loads an instance from the DB into the cache, just for the sake of deleting it.

Maybe `em.remove(em.getReference(cls, pk))` could be used to avoid loading the instance, but it may also just load the full instance at a later time although only the PK is needed and known beforehand. Anyway, it should be simpler than that.

**Comment by gavinking** [comment]
> Maybe `em.remove(em.getReference(cls, pk))` could be used to avoid loading the instance,

Indeed, exactly this is already supported in Hibernate 6, and I think it's a fine solution.

> Anyway, it should be simpler than that.

Really, I think that's quite simple enough.

**Comment by gavinking** [comment]
I'm going to go ahead and close this one, because the following is already possible:

    em.remove(session.getReference(Publisher.class, 1));

In general not every case is optimizable, because of the need to process cascades, etc. But it's for precisely this reason that `delete(id)` would have very weak semantics, and why we've never added it in all these years.

For those who really want this handed to them on a plate, Jakarta Data has a reasonable solution.
